STAN.CLIENT/StanConnection.cs

```diff
@@ -234,6 +267,207 @@ internal Connection(string stanClusterID, string clientID, StanOptions options)
             ackSubscription.SetPendingLimits(1024 * 1024, 32 * 1024 * 1024);
 
             pubAckMap = new BlockingDictionary<string, PublishAck>(opts.maxPubAcksInflight);
+
+            // TODO - check out sub map and chans
+
+            bool unsubPing = true;
+
+            // Do this with servers which are at least at protcolOne.
+            if (response.Protocol >= StanConsts.protocolOne)
+            {
+
+                // Note that in the future server may override client ping
+                // interval value sent in ConnectRequest, so use the
+                // value in ConnectResponse to decide if we send PINGs
+                // and at what interval.
+                // In tests, the interval could be negative to indicate
+                // milliseconds.
+                if (response.PingInterval != 0) {
+                    unsubPing = false;
+
+                    // These will be immutable
+                    pingRequests = response.PingRequests;
+                    pingInbox = pingSubscription.Subject;
+
+                    // negative values returned from the server are ms
+                    if (response.PingInterval < 0)
+                    {
+                        pingInterval = response.PingInterval * -1;
+                    }
+                    else
+                    {
+                        // if positive, the result is in seconds, but 
+                        // in the .NET clients we always use millis.
+                        pingInterval = response.PingInterval * 1000;
+                    }
+
+                    pingMaxOut = response.PingMaxOut;
+                    pingBytes = ProtocolSerializer.createPing(connID);
+
+                    lock (pingLock)
+                    {
+                        pingTimer = new Timer(pingServer, null, pingInterval, Timeout.Infinite);
+                    }
+                }
+            }
+            if (unsubPing)
+            {
+                pingSubscription.Unsubscribe();

```

Comment:
it looks like a few things above (e.g. ConnectRequest work) can cause an exception to be throw after pingSubscription is created on line 202.